File: demo/realtime-api/domain/models/Vehicle.js
import mongoose from 'mongoose';

const VehicleSchema = new mongoose.Schema({
  line: { type: mongoose.Schema.Types.ObjectId, ref: 'Line', required: true },
  mode: { type: String, enum: ['bus', 'tram', 'train', 'metro', 'ship', 'ferry', 'taxi'], required: true },
  status: { type: String, enum: ['running', 'stopped'], default: 'running' },
  currentStopIndex: { type: Number, default: 0 },
  progressBetweenStops: { type: Number, default: 0 }, // 0..1
  direction: { type: Number, enum: [0,1], default: 0 },
  lastUpdateTs: { type: Date, default: Date.now },
  position: {
    type: { type: String, enum: ['Point'], default: 'Point' },
    coordinates: { type: [Number], default: [0,0] },
  },
}, { timestamps: true });

VehicleSchema.index({ line: 1 });

export const Vehicle = mongoose.model('Vehicle', VehicleSchema);

File: shared/realtime/src/commonMain/kotlin/com/umain/transport/realtime/domain/model/Vehicle.kt
package com.umain.transport.realtime.domain.model

import kotlin.js.ExperimentalJsExport
import kotlin.js.JsExport

/**
 * A bus, train, tram, or boat actively running on a line — either being
 * advanced by the server's SimulationEngine or by a future on-device tick
 * loop (the offline-trips angle the talk hints at).
 *
 * Position is a [longitude, latitude] pair as floats so JS consumers can
 * pipe it straight into Mapbox/Leaflet without intermediate parsing.
 *
 * The fields mirror the realtime simulator's Mongoose Vehicle schema, but
 * stripped to the JS-friendly subset — no Mongo ObjectIds, no nested
 * GeoJSON wrappers, no cross-collection refs.
 */
@OptIn(ExperimentalJsExport::class)
@JsExport
data class Vehicle(
    val id: String,
    val lineId: String,
    val mode: String,
    val status: String,
    val currentStopIndex: Int,
    val progressBetweenStops: Double,
    val direction: Int,
    val longitude: Double,
    val latitude: Double,
    /**
     * Epoch ms as Double — Kotlin/JS in 2.3 can't @JsExport `Long` without
     * BigInt support, and BigInt breaks `JSON.stringify` for the same
     * reason `Line.transportMode` is a String instead of a Kotlin enum.
     * Double has 53 bits of integer precision, so timestamps stay accurate
     * past the year 287,000.
     */
    val lastUpdateMillis: Double,
)
